src/mame/drivers/firebeat.cpp

```diff
@@ -218,6 +224,7 @@ class firebeat_state : public driver_device
 	optional_device_array<k057714_device, 2> m_gcu;
 	optional_device<cy7c131_device> m_dpram;
 	optional_device<ata_interface_device> m_spuata;
+	optional_shared_ptr<uint16_t> m_waveram;
```

Comment:
There are a lot of optional devices here – would it be better to split this state class up by machine variant?

Reply:
Every game in the Firebeat driver has its own unique machine configuration which lead to this long list of optionals. I think refactoring the driver to split up the class for each variant might be better. Now that everything is actually working and in place for the most part that should hopefully be easier.

I'll see what I can do to clean things up but I don't want to commit to a full refactor for this PR.

Reply:
https://github.com/987123879113/mame/blob/firebeat_refactor/src/mame/drivers/firebeat.cpp

I made a rough attempt at refactoring the driver. I may have gone a little heavy. I split monolith `firebeat_state` into two base classes from which each game's machine is derived from: 
`class firebeat_state : public driver_device`
`class firebeat_spu_state : public firebeat_state` 

Each machine configuration has its own state:
`class firebeat_ppp_state : public firebeat_state`
`class firebeat_kbm_state : public firebeat_state` (no SPU but it has a bunch of MIDI keyboard UART stuff and also a sub GCU chip for the second monitor)
`class firebeat_bm3_state : public firebeat_spu_state` (configured for HDD, will also eventually have a FDD)
`class firebeat_popn_state : public firebeat_spu_state` (configured for CD-ROM)

I also moved any machine-specific code like MIDI keyboards for kbm, lights for a specific game, etc into their respective state class. This gets rid of all of the optional devices that aren't used between each game and hopefully organizes things a little better.

It's a pretty big modification with 1,109 additions and 1,010 deletions so I would prefer this to be its own PR if these changes sound good to you.

Reply:
Yeah, sounds good.  It can wait for another PR.